### PR TITLE
#20 resolve action links in excal app and update url action

### PR DESCRIPTION
**Pull request by asmaNode**
This PR must be merged before : https://github.com/OpenPaaS-Suite/esn-frontend-calendar/pull/170
**DEMO** :

updatelinkEXCAL

### CODE REVIEW

File: src/app/event-external-root/event-external-root.controller.js [hunk 1]
```diff
@@ -3,14 +3,29 @@
 angular.module('esn.calendar.event-consultation')
   .controller('CalEventExternalRootController', CalEventExternalRootController);
 
-function CalEventExternalRootController($scope, result) {
+function CalEventExternalRootController($rootScope, $location, result, calEventExternalRootService) {
   const self = this;
 
+  $rootScope.$on('UPDATE_ACTION_EXCAL', function(event, newLink) {
+    const url = new URL('http://localhost:8080' + newLink);
```
Comment (MichaelBailly) [inline src/app/event-external-root/event-external-root.controller.js:10]: I'm pretty sure http://localhost:8080/ is not the correct URL.

Comment (asmaNode) [reply src/app/event-external-root/event-external-root.controller.js:10]: the goal of this line is to create a fake url to extract jwt params but i am still working on it

File: src/app/event-external-root/event-external-root.controller.js [hunk 2]
```diff
@@ -3,14 +3,37 @@
...
+function CalEventExternalRootController($rootScope, $location, result, calEventExternalRootService) {
   const self = this;
 
+  $rootScope.$on('UPDATE_ACTION_EXCAL', function(event, newLink) {
+    const newJWT = parseJwtKey(newLink);
+
+    if (newJWT) {
+      $location.search('jwt', newJWT);
+    }
+
```
Comment (rezk2ll) [inline src/app/event-external-root/event-external-root.controller.js:15]:
```suggestion
```

File: src/app/event-external-root/event-external-root.controller.js [hunk 3]
```diff
@@ -3,14 +3,37 @@
...
     self.attendeeEmail = result.attendeeEmail;
-    self.links = result.links;
+    self.links = {};
+    Object.keys(result.links).forEach(action => {
+      const url = new URL(result.links[action]);
+      const jwt = url.searchParams.get('jwt');
+
+      self.links[action] = calEventExternalRootService.getChangeParticipationUrl(jwt);
+    });
+
```
Comment (rezk2ll) [inline src/app/event-external-root/event-external-root.controller.js:30]:
```suggestion
```

File: src/app/event-external-root/event-external-root.controller.js [hunk 4]
```diff
@@ -3,14 +3,34 @@
 angular.module('esn.calendar.event-consultation')
   .controller('CalEventExternalRootController', CalEventExternalRootController);
 
-function CalEventExternalRootController($scope, result) {
+function CalEventExternalRootController($rootScope, $location, result, calEventExternalRootService) {
   const self = this;
 
+  $rootScope.$on('UPDATE_ACTION_EXCAL', function(event, newLink) {
+    const newJWT = parseJwtKey(newLink);
```
Comment (renaudboyer) [inline src/app/event-external-root/event-external-root.controller.js:10]: If UPDATE_ACTION_EXCAL is only used to update the JWT, we can emit only the JWT if present in the link.

File: src/app/event-external-root/event-external-root.controller.js [hunk 5]
```diff
@@ -3,14 +3,34 @@
 angular.module('esn.calendar.event-consultation')
   .controller('CalEventExternalRootController', CalEventExternalRootController);
 
-function CalEventExternalRootController($scope, result) {
+function CalEventExternalRootController($rootScope, $location, result, calEventExternalRootService) {
   const self = this;
 
+  $rootScope.$on('UPDATE_ACTION_EXCAL', function(event, newLink) {
```
Comment (huy-ta) [inline src/app/event-external-root/event-external-root.controller.js:9]: Have a const for `UPDATE_ACTION_EXCAL`.

File: src/app/event-external-root/event-external-root.controller.js [hunk 6]
```diff
@@ -3,14 +3,34 @@
...
+      const jwt = url.searchParams.get('jwt');
+
+      self.links[action] = calEventExternalRootService.getChangeParticipationUrl(jwt);
+    });
+  }
+
+  function parseJwtKey(link) {
+    const token = link.match(/jwt=(.*)/);
+
+    return token ? token[1] : null;
```
Comment (huy-ta) [inline src/app/event-external-root/event-external-root.controller.js:34]: This is essentially doing the same thing as this:

```js
const url = new URL(result.links[action]);
const jwt = url.searchParams.get('jwt');
```

So you need to decide which one to use and put them in `parseJwt`.

File: src/app/event-external-root/event-external-root.controller.js [hunk 7]
```diff
@@ -3,14 +3,34 @@
...
   if (result.error) {
     self.error = result.error;
   } else {
     self.eventJcal = result.eventJSON;
     self.attendeeEmail = result.attendeeEmail;
-    self.links = result.links;
+    self.links = {};
+    Object.keys(result.links).forEach(action => {
+      const url = new URL(result.links[action]);
+      const jwt = url.searchParams.get('jwt');
```
Comment (huy-ta) [inline src/app/event-external-root/event-external-root.controller.js:25]:
```suggestion
      const jwt = parseJwt(link);
```

File: src/app/event-external-root/event-external-root.controller.js [hunk 8]
```diff
@@ -3,14 +3,34 @@
...
+
+      self.links[action] = calEventExternalRootService.getChangeParticipationUrl(jwt);
+    });
+  }
+
+  function parseJwtKey(link) {
```
Comment (huy-ta) [inline src/app/event-external-root/event-external-root.controller.js:31]:
```suggestion
  function parseJwt(link) {
```

**Review by huy-ta — COMMENTED** [review]
The flow looks good to me. Now you need to write tests for the changes.

File: src/app/event-external-root/event-external-root.controller.js (hunk 6 above)
Comment (asmaNode) [reply src/app/event-external-root/event-external-root.controller.js:34]: const url = new URL(result.links[action]); it is used for a valid URL and this is not the case for the link sent from calendar 
so we will use parseJwtKey

File: src/app/event-external-root/event-external-root.controller.js (hunk 4 above)
Comment (asmaNode) [reply src/app/event-external-root/event-external-root.controller.js]: we need to parse jwt key anyway (before emit or after)

**Review by rezk2ll — DISMISSED** [review]

File: src/app/event-external-root/event-external-root.controller.spec.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+
+/* global chai, sinon: false */
+
+var expect = chai.expect;
+
+describe.only('The CalEventExternalRootController controller', function() {
```
Comment (huy-ta) [inline src/app/event-external-root/event-external-root.controller.spec.js:7]:
```suggestion
describe('The CalEventExternalRootController controller', function() {
```

File: src/app/event-external-root/event-external-root.controller.spec.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+
+/* global chai, sinon: false */
+
+var expect = chai.expect;
```
Comment (huy-ta) [inline src/app/event-external-root/event-external-root.controller.spec.js:5]:
```suggestion
const { expect } = chai;
```

File: src/app/event-external-root/event-external-root.controller.spec.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+describe.only('The CalEventExternalRootController controller', function() {
+  var $controller, $rootScope, result, calEventExternalRootServiceMock, locationMock, $location, CAL_EVENTS;
```
Comment (huy-ta) [inline src/app/event-external-root/event-external-root.controller.spec.js:8]:
```suggestion
  let $controller, $rootScope, result, calEventExternalRootServiceMock, locationMock, $location, CAL_EVENTS;
```

File: src/app/event-external-root/event-external-root.controller.spec.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    });
+  });
+
+  function initController() {
+    return $controller('CalEventExternalRootController');
+  }
+
+  it('should change url jwt param when the event UPDATE_ACTION_EXCAL is received', function() {
+    initController();
+    var link = 'http://localhost:8888/test?jwt=123456';
```
Comment (huy-ta) [inline src/app/event-external-root/event-external-root.controller.spec.js:46]:
```suggestion
    const link = 'http://localhost:8888/test?jwt=123456';
```

File: src/app/event-external-root/event-external-root.controller.spec.js [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+    $location.search = sinon.spy();
+    $rootScope.$emit(CAL_EVENTS.UPDATE_ACTION_EXCAL, link);
+    expect($location.search).to.have.been.calledWith('jwt', '123456');
+  });
+
```
Comment (huy-ta) [inline src/app/event-external-root/event-external-root.controller.spec.js:52]:
```suggestion
```

File: src/app/event-external-root/event-external-root.controller.js [hunk 14]
```diff
@@ -3,14 +3,33 @@
 angular.module('esn.calendar.event-consultation')
   .controller('CalEventExternalRootController', CalEventExternalRootController);
 
-function CalEventExternalRootController($scope, result) {
+function CalEventExternalRootController($rootScope, $location, result, CAL_EVENTS, calEventExternalRootService) {
   const self = this;
 
+  $rootScope.$on(CAL_EVENTS.UPDATE_ACTION_EXCAL, function(event, newLink) {
+    const newJWT = parseJwt(newLink);
```
Comment (huy-ta) [inline src/app/event-external-root/event-external-root.controller.js:10]:
```suggestion
    const newJwt = parseJwt(newLink);
```

**Review by renaudboyer — APPROVED** [review]

**Review by rezk2ll — APPROVED** [review]

**Pull request merged**